client/core/libxc/binance.go

```diff
@@ -0,0 +1,1156 @@
+		if err != nil {
+			return nil, nil, err
+		}
+
+		cm := dex.NewConnectionMaster(conn)
+		if err = cm.ConnectOnce(bnc.ctx); err != nil {
+			return nil, nil, fmt.Errorf("websocketHandler remote connect: %v", err)
+		}
+
+		return conn, cm, nil
+	}
+
+	conn, cm, err := newConn()
+	if err != nil {
+		return fmt.Errorf("error initializing connection: %v", err)
+	}
+	bnc.marketStream = conn
+
+	bnc.wg.Add(1)
+	go func() {
+		defer bnc.wg.Done()
+		defer func() {
+			bnc.storeMarketStream(nil)
+			if cm != nil {
+				cm.Disconnect()
+			}
+		}()
+
+		reconnect := time.After(time.Hour * 12)
+		for {
+			select {
+			case <-reconnect:
+				if cm != nil {
+					cm.Disconnect()
+				}
+				conn, cm, err = newConn()
+				if err != nil {
+					bnc.log.Errorf("error reconnecting: %v", err)
+					reconnect = time.After(time.Second * 30)
+				} else {
+					bnc.storeMarketStream(conn)
+					reconnect = time.After(time.Hour * 12)
+				}
+			case <-bnc.ctx.Done():
+				return
+			}
+		}
+	}()
+
+	return nil
+}
+
+// UnsubscribeMarket unsubscribes from order book updates on a market.
+func (bnc *binance) UnsubscribeMarket(baseSymbol, quoteSymbol string) error {
+	return bnc.stopMarketDataStream(strings.ToLower(baseSymbol + quoteSymbol))
+}
+
+// SubscribeMarket subscribes to order book updates on a market. This must
+// be called before calling VWAP.
+func (bnc *binance) SubscribeMarket(baseSymbol, quoteSymbol string) error {
+	return bnc.startMarketDataStream(baseSymbol, quoteSymbol)
+}
+
+// VWAP returns the volume weighted average price for a certain quantity
+// of the base asset on a market.
+func (bnc *binance) VWAP(baseSymbol, quoteSymbol string, sell bool, qty uint64) (avgPrice, extrema uint64, filled bool, err error) {
+	fail := func(err error) (uint64, uint64, bool, error) {
+		return 0, 0, false, err
+	}
+
+	slug := strings.ToLower(baseSymbol + quoteSymbol)
+	var side []*bookBin
+	bnc.booksMtx.RLock()
+	book, found := bnc.books[slug]
+	if found {
+		if sell {
+			side = book.asks
+		} else {
+			side = book.bids
+		}
+	}
+	bnc.booksMtx.RUnlock()
+	if side == nil {
+		return fail(fmt.Errorf("no book found for %s", slug))
+	}
+
+	baseCfg, err := bncSymbolData(baseSymbol)
+	if err != nil {
+		return fail(fmt.Errorf("error getting symbol data for %s: %w", baseSymbol, err))
+	}
+
+	quoteCfg, err := bncSymbolData(quoteSymbol)
+	if err != nil {
+		return fail(fmt.Errorf("error getting symbol data for %s: %w", quoteSymbol, err))
+	}
+
+	remaining := qty
+	var weightedSum uint64
+	for _, bin := range side {
+		extrema = calc.MessageRateAlt(bin.Price, baseCfg.conversionFactor, quoteCfg.conversionFactor)
+		binQty := uint64(bin.Qty * float64(baseCfg.conversionFactor))
+		if binQty >= remaining {
+			filled = true
+			weightedSum += remaining * extrema
+			break
+		}
+		remaining -= binQty
+		weightedSum += binQty * extrema
+	}
+
+	if !filled {
+		return 0, 0, false, nil
+	}
+
+	return weightedSum / qty, extrema, true, nil
+}
+
+type binanceNetworkInfo struct {
```

Comment:
Can you comment a url where the api specifics exist in case something ever changes?